libvast/test/printer_plugin.cpp

```diff
@@ -0,0 +1,336 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "vast/collect.hpp"
+#include "vast/plugin.hpp"
+#include "vast/table_slice.hpp"
+#include "vast/table_slice_builder.hpp"
+#include "vast/test/test.hpp"
+
+#include <caf/test/dsl.hpp>
+using namespace vast;
+
+namespace {
+
+// Builds a chain of events where consecutive chunks of
+// num_events_per_type events have the same type.
+struct basic_table_slice_generator {
+  id offset;
+
+  explicit basic_table_slice_generator(type input_schema)
+    : offset(0), schema(input_schema) {
+  }
+
+  table_slice operator()(size_t num) {
+    auto builder = std::make_shared<table_slice_builder>(schema);
+    for (size_t i = 0; i < num; ++i) {
+      CHECK(builder->add(make_data_view("foo")));
+    }
+    auto slice = builder->finish();
+    slice.offset(offset);
+    offset += num;
+    return slice;
+  }
+
+  type schema;
+};
+
+struct fixture {
+  struct mock_control_plane : operator_control_plane {
+    [[nodiscard]] virtual auto self() noexcept -> caf::event_based_actor& {
+      FAIL("no mock implementation available");
+    }
+    virtual auto stop([[maybe_unused]] caf::error error = {}) noexcept -> void {
+      FAIL("no mock implementation available");
+    }
+
+    virtual auto warn([[maybe_unused]] caf::error warning) noexcept -> void {
+      FAIL("no mock implementation available");
+    }
+
+    virtual auto emit([[maybe_unused]] table_slice metrics) noexcept -> void {
+      FAIL("no mock implementation available");
+    }
+
+    [[nodiscard]] virtual auto
+    demand([[maybe_unused]] type schema = {}) const noexcept -> size_t {
+      FAIL("no mock implementation available");
+    }
+
+    [[nodiscard]] virtual auto schemas() const noexcept
+      -> const std::vector<type>& {
+      FAIL("no mock implementation available");
+    }
+
+    [[nodiscard]] virtual auto concepts() const noexcept
+      -> const concepts_map& {
+      FAIL("no mock implementation available");
+    }
+  };
+
+  fixture() {
```

Comment:
Possible improvement: There's a `fixtures::events` fixture that contains a bunch of table slices that you could use for testing.

Reply:
Wow, `fixtures::events` is really handy. I added two unit tests using data from `fixtures::events` in https://github.com/tenzir/vast/pull/3009/commits/561a6e810c7481799fe338f5b7b4555e306d51a8